### PR TITLE
API: add path support (navigation)

### PR DESCRIPTION
**Pull request by ottodevs**
**Changes**
Added request handler to the wrapper, then handle path from there and finally connect API and slightly modify API-react to use this new handler from the observable

**If you are modifying the external API of one of the modules, please remember to also change the [documentation](/docs/)**
Pending docs until some approval feedback / polish

- [x] I have updated the associated documentation with my changes
TBD

### CODE REVIEW

File: packages/aragon-api-react/src/index.js [hunk 1]
```diff
@@ -81,6 +77,9 @@ function AragonApi({
 
           // network
           api.network().subscribe(network => setNetwork(network || null)),
+
+          // path
+          api.path().subscribe(path => setPath(path)),
```
Comment (sohkai) [inline packages/aragon-api-react/src/index.js]: Does a default value make sense here? Like a `''` or `'/'`?

File: packages/aragon-wrapper/src/rpc/handlers/path.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export default function (wrapper) {
+    return wrapper.path
+}
```
Comment (sohkai) [inline packages/aragon-wrapper/src/rpc/handlers/path.js]: It probably makes sense to use a similar pattern to the [`cache`](https://github.com/aragon/aragon.js/blob/master/packages/aragon-wrapper/src/rpc/handlers/cache.js#L3) handler here (`get` vs. `request` path).

File: packages/aragon-api/src/index.js [hunk 3]
```diff
@@ -61,6 +61,19 @@ export class AppProxy {
...
+   * @return {Observable} Multi-emission Observable that emits a path every time a change is detected.
+   */
+  path () {
+    return this.rpc.sendAndObserveResponses(
+      'path'
+    ).pipe(
+      pluck('result')
+    )
+  }
+
```
Comment (sohkai) [inline packages/aragon-api/src/index.js]: Reminder to add `requestPath()` here and in `api-react` :)

File: packages/aragon-wrapper/src/index.js [hunk 4]
```diff
@@ -988,6 +989,27 @@ export default class Aragon {
...
   }
 
+  /**
+   * Initialise the path observable.
+   *
+   * @return {Promise<void>}
+   */
+  async initPath () {
+    this.path = new ReplaySubject(1)
+    this.path.next('/')
```
Comment (sohkai) [inline packages/aragon-wrapper/src/index.js]: This observable is the only one that I'd think a bit more on, and be careful with. A couple of thoughts:
- Having `this.path` transparently represent the current app isn't safe if a script wanted to try and use `requestPath()`.
- What we probably want is a request/response scheme, where there's an `this.pathIntents` (for sake of consistency with other "intents") that the client receives requests on, and it either accepts or rejects those intents.
- If a path change intent is accepted, we push a change to `this.path` with the proxy address and the accepted path. At any time, the client can use `setPath(proxyAddress, path)` to change the path for an app (useful, for example, when loading the org at `<org>/<voting address>/vote/3`; the client should immediately tell the voting app to load the vote once the handshake is complete).
- The `path` handler should filter out `this.path` emissions that don't match its own proxy address

**Comment by sohkai** [comment]
@ottodevs I've done a few things to this PR:

- Rebased it against master
- Cherry-picked `api-react` changes to that PR (#326) 
- Added request path APIs
- Updated the implementation based on https://github.com/aragon/aragon.js/pull/352#discussion_r307400300

In particular, clients should now respond to `pathIntent`s and set the corresponding app's path using `wrapper.setAppPath()` when allowing a request.

File: packages/aragon-api/src/index.js [hunk 5]
```diff
@@ -139,6 +126,34 @@ export class AppProxy {
     )
   }
 
+  /**
+   * Get current path for the app. Useful for in-app routing and navigation.
+   *
+   * @return {Observable} Multi-emission Observable that emits the app's current path every time a change is detected.
+   */
+  paths () {
```
Comment (bpierre) [inline packages/aragon-api/src/index.js]: Shouldn’t it be named `path`? Even if it updates, it’s always a single path that is being received.

Comment (sohkai) [reply packages/aragon-api/src/index.js]: Hmm yes, we do name `network` as singular even though you're observing the network changes over time. Let's rename it back to `path()`.

`getApps()` should be named `apps` or `installedApps`, but is a bit problematic due to backwards compatibility issues. I'll add `installedApps` and deprecate `getApps()`.

**Comment by sohkai** [comment]
@bpierre I've also thought about if it made sense to pass along more than a string; perhaps we can pass a string and an optional object that contains client-specific behaviour.

Ideally asking to navigate to something like the email notifications would be more robust than harding coding the client's preferences query string into the path request.

But we can do this on a second iteration since this change would be backwards compatible with the current API.

File: packages/aragon-api/src/index.js (hunk 5 above)
Comment (sohkai) [reply packages/aragon-api/src/index.js]: Changes for `installedApps()` and `currentPath()` are in https://github.com/aragon/aragon.js/pull/374.

`paths()` has been renamed to `path()` in 1e251f1.

**Pull request merged**